src/plugins/preauth/pkinit/pkinit_srv.c

```diff
@@ -1489,3 +1567,284 @@ kdcpreauth_pkinit_initvt(krb5_context context, int maj_ver, int min_ver,
     vt->return_padata = pkinit_server_return_padata;
     return 0;
 }
+
+/*
+ * Run the received, verified certificate through the certauth plugin interface,
+ * to verify that client can use the certificate based on the extension policy
+ * defined by modules.
+ */
+static krb5_error_code
+server_authorize_cert(krb5_context context, certauth_handle *certauth_modules,
+                      pkinit_kdc_context plgctx,
+                      pkinit_kdc_req_context reqctx,
+                      krb5_kdcpreauth_callbacks cb, krb5_kdcpreauth_rock rock,
+                      krb5_principal client)
+{
+    krb5_error_code ret;
+    certauth_handle hd;
+    struct certauth_req_opts opts;
+    krb5_boolean accepted = FALSE;
+    krb5_data cert;
+    size_t i;
+    void *db_ent = NULL;
+    char **ais = NULL, **ai = NULL;
+
+    /* Re-encode the received certificate into DER, which is extra work, but
+     * avoids creating a crypto dependency on the interface. */
+    ret = crypto_encode_der_cert(context, reqctx->cryptoctx, &cert);
+    if (ret)
+        goto cleanup;
+
+    /* Set options for the builtin module. */
+    opts.plgctx = plgctx;
+    opts.reqctx = reqctx;
+    opts.cb = cb;
+    opts.rock = rock;
+
+    db_ent = cb->client_entry(context, rock);
+
+    /*
+     * Check the certificate against each certauth module. For the certificate
+     * to be authorized at least one module must return 0 and no module returns
+     * an error code other than KRB5_PLUGIN_NO_HANDLE (pass).  Add indicators
+     * from modules that return 0 or pass.
+     */
+    ret = KRB5_PLUGIN_NO_HANDLE;
+    for (i = 0; certauth_modules != NULL && certauth_modules[i] != NULL; i++) {
+        hd = certauth_modules[i];
+        if (hd->vt.authorize == NULL)
+            continue;
+
+        ret = hd->vt.authorize(context, hd->moddata, cert, client, &opts,
+                               db_ent, &ais);
+
+        if (ret == 0)
+            accepted = TRUE;
+        else if (ret != KRB5_PLUGIN_NO_HANDLE)
+            goto cleanup;
+
+        if (ais != NULL) {
+            /* Assert authentication indicators from certificate authorization. */
+            for (ai = ais; *ai != NULL; ai++) {
+                ret = cb->add_auth_indicator(context, rock, *ai);
+                if (ret)
+                    goto cleanup;
+            }
+
+            if (hd->vt.free_ind != NULL)
+                hd->vt.free_ind(context, hd->moddata, ais);
+            else
+                free_ai_list(ais);
+            ais = NULL;
+        }
+    }
+
+    if (accepted)
+        ret = 0;
+
+cleanup:
+    crypto_free_cert_data(context, cert);
+    return ret;
+}
+
+static krb5_error_code
+pkinitsrv_authorize(krb5_context context, krb5_certauth_moddata moddata,
+                   krb5_data cert, krb5_principal princ, void *opts,
+                   void *db_entry, char ***authinds_out)
```

Comment:
I don't think we're feeding the right things into the accumulator here, with the result that administrators will have to use module configuration to disable the pkinitsrv module when they shouldn't have to.

If the EKU is not found, I think we should say no (KRB5KDC_ERR_INCONSISTENT_KEY_PURPOSE).  An admin can disable EKU checking in configuration if desired.

If the SAN is not found, I think we should pass.  The certificate may still match the client principal according to another module's criteria, and there is no way to disable SAN checking in configuration.